### PR TITLE
Fixes #14787 building on #14811:  Set Origin header in http_network_or_cache_fetch

### LINKED ISSUE #14787: Set origin header in http_network_or_cache_fetch
There's a TODO comment pointing at https://github.com/hyperium/hyper/pull/691, but that was closed in favour of https://github.com/hyperium/hyper/commit/01843f882265a894c7051dc2ecf5cf09f2c2e8e7. We should be able to address the TODO now.

Code: `components/net/http_loader.rs`

### PR DESCRIPTION
**Pull request by samgiles**
Incorporates review feedback from #14811 

I noticed that the "steps" in the comments don't necessarily correspond to the steps in the current fetch standard.  This adds step 11 in the spec, but the steps commented in the implementation look wrong.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #14787.
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15547)

### CODE REVIEW

**Comment by Wafflespeanut** [comment]
Please comment on the issue before working on it, so that we can avoid duplicate efforts.

**Comment by Ms2ger** [comment]
@bors-servo try

Unit test is great, but we should also have a wpt test. Let's see if one exists.

**Comment by Ms2ger** [comment]
```

Tests with unexpected results:
  ▶ Unexpected subtest result in /XMLHttpRequest/open-url-worker-origin.htm:
  └ PASS [expected FAIL] Origin header

  ▶ TIMEOUT [expected OK] /XMLHttpRequest/send-entity-body-document.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ Unexpected subtest result in /cors/allow-headers.htm:
  └ PASS [expected FAIL] Allow origin: *

  ▶ Unexpected subtest result in /cors/allow-headers.htm:
  └ PASS [expected FAIL] Allow origin: _*__

  ▶ Unexpected subtest result in /cors/allow-headers.htm:
  └ PASS [expected FAIL] Allow origin: [tab]*

  ▶ Unexpected subtest result in /cors/allow-headers.htm:
  └ PASS [expected FAIL] Allow origin: http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/allow-headers.htm:
  └ PASS [expected FAIL] Allow origin: _http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/allow-headers.htm:
  └ PASS [expected FAIL] Allow origin: [tab]http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/basic.htm:
  └ PASS [expected FAIL] Cross domain basic usage

  ▶ Unexpected subtest result in /cors/basic.htm:
  └ PASS [expected FAIL] Same domain different port

  ▶ Unexpected subtest result in /cors/basic.htm:
  └ PASS [expected FAIL] Cross domain different port

  ▶ OK [expected TIMEOUT] /cors/credentials-flag.htm

  ▶ Unexpected subtest result in /cors/credentials-flag.htm:
  └ PASS [expected TIMEOUT] Don't send cookie by default

  ▶ Unexpected subtest result in /cors/credentials-flag.htm:
  └ PASS [expected TIMEOUT] Don't send cookie part 2

  ▶ Unexpected subtest result in /cors/credentials-flag.htm:
  └ PASS [expected TIMEOUT] Don't obey Set-Cookie when withCredentials=false

  ▶ Unexpected subtest result in /cors/credentials-flag.htm:
  │ FAIL [expected PASS] Access-Control-Allow-Credentials: TRUE should be disallowed (async)
  │   → assert_unreached: onload Reached unreachable code
  │ FAIL [expected PASS] Access-Control-Allow-Credentials: True should be disallowed (async)
  │   → assert_unreached: onload Reached unreachable code
  │ 
  │ test_response_header/</client.onload<@http://web-platform.test:8000/cors/credentials-flag.htm:112:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1433:20

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  └ PASS [expected FAIL] local (*) to remote (*), expect origin=http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  └ PASS [expected FAIL] local (*) to remote (http://web-platform.test:8000), expect origin=http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  └ PASS [expected FAIL] local (http://web-platform.test:8000) to remote (*), expect origin=http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  └ PASS [expected FAIL] local (http://web-platform.test:8000) to remote (http://web-platform.test:8000), expect origin=http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  └ PASS [expected FAIL] local (null) to remote (*), expect origin=http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  └ PASS [expected FAIL] local (none) to remote (*), expect origin=http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/origin.htm:
  └ PASS [expected FAIL] Allow origin: *

  ▶ Unexpected subtest result in /cors/origin.htm:
  └ PASS [expected FAIL] Allow origin: _*__

  ▶ Unexpected subtest result in /cors/origin.htm:
  └ PASS [expected FAIL] Allow origin: [tab]*

  ▶ Unexpected subtest result in /cors/origin.htm:
  └ PASS [expected FAIL] Allow origin: http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/origin.htm:
  └ PASS [expected FAIL] Allow origin: _http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/origin.htm:
  └ PASS [expected FAIL] Allow origin: [tab]http://web-platform.test:8000

  ▶ Unexpected subtest result in /cors/request-headers.htm:
  └ PASS [expected FAIL] basic request header

  ▶ Unexpected subtest result in /cors/request-headers.htm:
  └ PASS [expected FAIL] Simple request headers need not be in allow-headers

  ▶ Unexpected subtest result in /cors/request-headers.htm:
  └ PASS [expected FAIL] Strange allowheaders (case insensitive)

  ▶ Unexpected subtest result in /cors/request-headers.htm:
  └ PASS [expected FAIL] INVALID_STATE_ERR on setRequestHeader after send()

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 200

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 201

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 202

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 203

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 204

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 205

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 206

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 209

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 299

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on POST 200

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on HEAD 200

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on PUT 200

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on CHICKEN 200

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on GET 400

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on HEAD 401

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on POST 404

  ▶ Unexpected subtest result in /cors/status-async.htm:
  └ PASS [expected FAIL] Status on POST 500

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on GET 200

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on GET 204

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on GET 400

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on GET 401

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on HEAD 200

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on HEAD 204

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on HEAD 400

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on HEAD 401

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on HEAD 501

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on HEAD 699

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on POST 204

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on POST 400

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on POST 401

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on POST 404

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on PUT 699

  ▶ Unexpected subtest result in /cors/status-preflight.htm:
  └ PASS [expected FAIL] CORS - status after preflight on CHICKEN 501

  ▶ Unexpected subtest result in /cors/status.htm:
  └ PASS [expected FAIL] 3. CORS allowed, response status 400.

  ▶ Unexpected subtest result in /cors/status.htm:
  └ PASS [expected FAIL] 4. CORS allowed, preflight status 200, response status 400.

  ▶ OK [expected TIMEOUT] /eventsource/request-cache-control.htm

  ▶ Unexpected subtest result in /eventsource/request-cache-control.htm:
  └ PASS [expected TIMEOUT] EventSource: Cache-Control 1

  ▶ OK [expected TIMEOUT] /eventsource/request-credentials.htm

  ▶ Unexpected subtest result in /eventsource/request-credentials.htm:
  └ PASS [expected TIMEOUT] EventSource: credentials: credentials enabled

  ▶ Unexpected subtest result in /eventsource/request-credentials.htm:
  └ PASS [expected TIMEOUT] EventSource: credentials: credentials disabled

  ▶ Unexpected subtest result in /eventsource/request-credentials.htm:
  └ PASS [expected TIMEOUT] EventSource: credentials: credentials default

  ▶ Unexpected subtest result in /fetch/api/basic/request-headers.html:
  └ PASS [expected FAIL] Fetch with PUT without body

  ▶ Unexpected subtest result in /fetch/api/basic/request-headers.html:
  └ PASS [expected FAIL] Fetch with POST without body

  ▶ Unexpected subtest result in /fetch/api/basic/request-headers.html:
  └ PASS [expected FAIL] Fetch with POST with Blob body

  ▶ Unexpected subtest result in /fetch/api/basic/request-headers.html:
  └ PASS [expected FAIL] Fetch with POST and mode "same-origin" needs an Origin header

  ▶ Unexpected subtest result in /fetch/api/basic/request-headers.html:
  └ PASS [expected FAIL] Fetch with POST and mode "no-cors" needs an Origin header

  ▶ Unexpected subtest result in /fetch/api/basic/request-headers.html:
  └ PASS [expected FAIL] Fetch with PUT and mode "same-origin" needs an Origin header

  ▶ Unexpected subtest result in /eventsource/eventsource-cross-origin.htm:
  └ PASS [expected TIMEOUT] EventSource: cross-origin basic use

  ▶ Unexpected subtest result in /eventsource/eventsource-cross-origin.htm:
  └ PASS [expected TIMEOUT] EventSource: cross-origin redirect use recon
```

The timeout in `send-entity-body-document.htm` is https://github.com/servo/servo/issues/15239.

It might be worth checking the failures in `credentials-flag.htm`, but I suspect those were tests that didn't run before.

Everything else looks excellent; please update the expectation files.

File: components/net/http_loader.rs [hunk 1]
```diff
@@ -789,6 +790,15 @@ fn http_redirect_fetch(request: Rc<Request>,
     main_fetch(request, cache, cors_flag, true, target, done_chan, context)
 }
 
+fn try_url_origin_to_hyper_origin(url_origin: &UrlOrigin) -> Option<HyperOrigin> {
+    match *url_origin {
+        // TODO Set "Origin: null" when hyper supports it
```
Comment (Ms2ger) [inline components/net/http_loader.rs]: File an issue on this.

**Comment by jdm** [comment]
Note: to update expectation files, see [the documentation](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations).

**Comment by samgiles** [comment]
It looks like the wpt failures are probably now related to #15569.

**Comment by samgiles** [comment]
Are new "unexpectedly passed" WPTs automatically updated by some process?

**Comment by jdm** [comment]
The same process by which expected failures are updated. If the entire file passes, then removing the ini file is enough. Otherwise, removing any `[test description] expected: FAIL` lines does the same job. If you follow the steps in the documentation, you can run the tests locally and update them automatically using `./mach update-wpt logfile`.

**Comment by samgiles** [comment]
/sgiles wonders if he can trigger a try

@bors-servo try

**Comment by Wafflespeanut** [comment]
Tests with unexpected results:
      ▶ OK [expected TIMEOUT] /XMLHttpRequest/send-network-error-async-events.sub.htm

      ▶ Unexpected subtest result in /XMLHttpRequest/send-network-error-async-events.sub.htm:
      └ PASS [expected TIMEOUT] XmlHttpRequest: The send() method: Fire a progress event named error when Network error happens (synchronous flag is unset)

      ▶ Unexpected subtest result in /cors/redirect-origin.htm:
      └ PASS [expected FAIL] local (*) to remote (*), expect origin=http://web-platform.test:8000

      ▶ Unexpected subtest result in /cors/redirect-origin.htm:
      └ PASS [expected FAIL] local (*) to remote (http://web-platform.test:8000), expect origin=http://web-platform.test:8000

      ▶ Unexpected subtest result in /cors/redirect-origin.htm:
      └ PASS [expected FAIL] local (http://web-platform.test:8000) to remote (*), expect origin=http://web-platform.test:8000

      ▶ Unexpected subtest result in /cors/redirect-origin.htm:
      └ PASS [expected FAIL] local (http://web-platform.test:8000) to remote (http://web-platform.test:8000), expect origin=http://web-platform.test:8000

      ▶ Unexpected subtest result in /cors/redirect-origin.htm:
      └ PASS [expected FAIL] local (null) to remote (*), expect origin=http://web-platform.test:8000

      ▶ Unexpected subtest result in /cors/redirect-origin.htm:
      └ PASS [expected FAIL] local (none) to remote (*), expect origin=http://web-platform.test:8000

      ▶ Unexpected subtest result in /eventsource/dedicated-worker/eventsource-constructor-non-same-origin.htm:
      │ FAIL [expected TIMEOUT] dedicated worker - EventSource: constructor (act as if there is a network error) (https://example.not/test)
      │   → assert_equals: source.readyState expected 2 but got 0
      │ 
      │ fetchFail/</worker.onmessage/<@http://web-platform.test:8000/eventsource/dedicated-worker/eventsource-constructor-non-same-origin.htm:31:15
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
      └ fetchFail/</worker.onmessage@http://web-platform.test:8000/eventsource/dedicated-worker/eventsource-constructor-non-same-origin.htm:29:13

      ▶ Unexpected subtest result in /eventsource/eventsource-constructor-non-same-origin.htm:
      │ FAIL [expected TIMEOUT] EventSource: constructor (act as if there is a network error) (https://example.not/test)
      │   → assert_equals: expected 2 but got 0
      │ 
      │ fetchFail/</source.onerror/<@http://web-platform.test:8000/eventsource/eventsource-constructor-non-same-origin.htm:18:15
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
      └ fetchFail/</source.onerror@http://web-platform.test:8000/eventsource/eventsource-constructor-non-same-origin.htm:17:13

**Comment by samgiles** [comment]
Eek, these seem different on Mac (I'm on Mac) :| [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2473).

**Comment by jdm** [comment]
If you look at the ini files for those tests, they're disabled on mac.

**Comment by samgiles** [comment]
Right, I think that's it.  

I was confused as I was expecting `./mach update-wpt [log file]` to update based on the log file, but it seems like it ignored `cors/redirect-origin.htm` because it's disabled on OSX.

**Comment by jdm** [comment]
@samgiles Does running `./mach test-wpt tests/wpt/web-platform-tests/fetch/nosniff/parsing-nosniff.html` show this failure for you?
```
  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: quoted-single
  └   → assert_unreached: Unexpected error event Reached unreachable code
```

**Comment by samgiles** [comment]
Right, pretty sure that's just https://github.com/servo/servo/issues/15239 now.

**Comment by samgiles** [comment]
@jdm I get that failure on master too, in fact, every time I run it, I get a different set of failures.

**Comment by jdm** [comment]
@bors-servo: retry
* #15631

File: tests/wpt/metadata/XMLHttpRequest/send-entity-body-document.htm.ini [hunk 2]
```diff
@@ -1,5 +1,6 @@
 [send-entity-body-document.htm]
   type: testharness
+  expected: TIMEOUT
```
Comment (jdm) [inline tests/wpt/metadata/XMLHttpRequest/send-entity-body-document.htm.ini]: This change should be reverted, since it's just a manifestation of #15239.

**Comment by KiChjang** [comment]
Looks like this needs some squashing now.

**Comment by Wafflespeanut** [comment]
Thanks!

**Comment by Wafflespeanut** [comment]
Looks like the `UrlOrigin` import has been removed in `master`. Please rebase your changes and ping me, and we'll land it :)

**Comment by nox** [comment]
@samgiles Are you still working on this?

**Comment by samgiles** [comment]
Been super busy, but will get a chance tomorrow evening to rebase this.

**Comment by ferjm** [comment]
@samgiles I tried to contact you via IRC. I needed these changes to fix #14519 in #15889.

**Comment by Wafflespeanut** [comment]
I tried pushing to this PR, but I couldn't. I've split the commits and opened the rebased version in #15903

**Pull request closed without merging**